### ISSUE TITLE
Parent doesn't switch to short address poll handling when the ack of the Child Id response got lost.

### ISSUE THREAD

**Issue by thcu-gp** [openthread/openthread#7613]
**Describe the bug** 
On indirect_sender.cpp:492 a scenario is described where the _MLE Child Id Response_ is received by the child, but the ACK not by the parent. However the callback `HandleSentFrameToChild` is not called by data_poll_handler.cpp:297 in case of a 1st no ack. Because of the `ExitNow` on line 286. Because the child did receive the _MLE Child Id Response_ correctly, it will never retry, the maxNumberOfRetries will never be reached and, `HandleSentFrameToChild` is never called and the parent will not make the switch from extended addresses to short addresses for the datapolls.

**To Reproduce** Information to reproduce the behavior, including:

1. 69bfc4b1c
2. qpg (but it shouldn't matter)
3. Leader build
4. 5.2.1 - Enhanced Frame Pending with Thread V1.2 End Device V1.1 End Device ( RD: 4 ) - Leader and SED (Thread1.2)

**Expected behavior**: The parent should switch to short address poll handling

**Console/log output** If applicable, add console/log output to help explain your problem.
```
2022-04-20-16:06:43.204 >> 728.696000 > [INFO]-MLE-----: Send Child ID Response (fe80:0:0:0:146e:a00:0:5,0xd803)
2022-04-20-16:06:43.218 >> 728.710000 > [DEBG]-MAC-----: Idle mode: Radio receiving on channel 11
2022-04-20-16:06:43.234 >> 728.726000 > [INFO]-CORE----: Notifier: StateChanged (0x00000400) [Child+]
2022-04-20-16:06:43.262 >> 728.754000 > THCU qorvoRadioAddSrcMatchExtEntry
2022-04-20-16:06:43.262 >> 728.754000 > THCU: ADD DP for nb: 166e0a00:00000005
2022-04-20-16:06:43.292 >> 728.784000 > [DEBG]-MAC-----: SrcAddrMatch - Adding addr: 166e0a0000000005 -- OK (0)
2022-04-20-16:06:43.324 >> 728.816000 > [DEBG]-MLE-----: THCU: Add Message (cnt=1) for Sleepy child with mleid[0-1]: BE6D, rloc16: D803
2022-04-20-16:06:43.339 >> 728.831000 > [INFO]-MAC-----: THCU: UpdateIndirectMessage
2022-04-20-16:06:43.355 >> 728.847000 > [INFO]-MAC-----: Prepping indir tx IPv6 UDP msg, len:96, chksum:7591, to:166e0a0000000005, sec:no, prio:net
2022-04-20-16:06:43.386 >> 728.878000 > [DEBG]-MAC-----: Rx security - frame counter 0
2022-04-20-16:06:43.401 >> 728.893000 > [INFO]-MAC-----: Rx data poll, src:0xd803, qed_msgs:1, rss:-53, ack-fp:1
2022-04-20-16:06:43.417 >> 728.909000 > [DEBG]-MAC-----: Request to start operation "TransmitDataIndirect"
2022-04-20-16:06:43.433 >> 728.925000 > [DEBG]-MAC-----: Starting operation "TransmitDataIndirect"
2022-04-20-16:06:43.449 >> 728.941000 > [DEBG]-MAC-----: RadioState: Receive -> CsmaBackoff
2022-04-20-16:06:43.464 >> 728.956000 > [DEBG]-MAC-----: RadioState: CsmaBackoff -> Transmit
2022-04-20-16:06:43.511 >> 729.003000 > [INFO]-MAC-----: Frame tx attempt 1/1 failed, error:NoAck, len:80, seqnum:146, type:Data, src:166e0a0000000001, dst:166e0a0000000005, sec:no, ackreq:
2022-04-20-16:06:43.527 >> 729.019000 > [DEBG]-MAC-----: RadioState: Transmit -> Receive
2022-04-20-16:06:43.543 >> 729.035000 > [DEBG]-MAC-----: Finishing operation "TransmitDataIndirect"
2022-04-20-16:06:43.558 >> 729.050000 > [INFO]-MAC-----: Indirect tx to child d803 failed, attempt 1/4
2022-04-20-16:06:43.574 >> 729.066000 > [INFO]-MAC-----: THCU: Max number retries not reached: err: 14
```
where the last line of logging was added before the aforementioned `ExitNow`

When the ack isn't lost, we get something along the lines of:
```
2022-04-20-16:06:35.277 >> 720.769000 > THCU qorvoRadioAddSrcMatchExtEntry
2022-04-20-16:06:35.308 >> 720.800000 > THCU: ADD DP for nb: 166e0a00:00000002
2022-04-20-16:06:35.324 >> 720.816000 > [DEBG]-MAC-----: SrcAddrMatch - Adding addr: 166e0a0000000002 -- OK (0)
2022-04-20-16:06:35.324 >> 720.816000 > [DEBG]-MAC-----: SrcAddrMatch - Enabling
2022-04-20-16:06:35.355 >> 720.847000 > [DEBG]-MLE-----: THCU: Add Message (cnt=1) for Sleepy child with mleid[0-1]: 7933, rloc16: D802
2022-04-20-16:06:35.386 >> 720.878000 > [INFO]-MAC-----: Prepping indir tx IPv6 UDP msg, len:96, chksum:d4b3, to:166e0a0000000002, sec:no, prio:net
2022-04-20-16:06:35.402 >> 720.894000 > [DEBG]-MAC-----: Rx security - frame counter 0
2022-04-20-16:06:35.418 >> 720.910000 > [INFO]-MAC-----: Rx data poll, src:0xd802, qed_msgs:1, rss:-52, ack-fp:1
2022-04-20-16:06:35.433 >> 720.925000 > [DEBG]-MAC-----: Request to start operation "TransmitDataIndirect"
2022-04-20-16:06:35.449 >> 720.941000 > [DEBG]-MAC-----: Starting operation "TransmitDataIndirect"
2022-04-20-16:06:35.464 >> 720.956000 > [DEBG]-MAC-----: RadioState: Receive -> CsmaBackoff
2022-04-20-16:06:35.480 >> 720.972000 > [DEBG]-MAC-----: RadioState: CsmaBackoff -> Transmit
2022-04-20-16:06:35.511 >> 721.003000 > [DEBG]-MAC-----: RadioState: Transmit -> Receive
2022-04-20-16:06:35.527 >> 721.019000 > [DEBG]-MAC-----: Finishing operation "TransmitDataIndirect"
2022-04-20-16:06:35.542 >> 721.034000 > [WARN]-MAC-----: THCU: HandleSentFrameToChild: enabled:1
2022-04-20-16:06:35.573 >> 721.065000 > [DEBG]-MLE-----: THCU: Set src match as short for Sleepy child with mleid[0-1]: 0001, rloc16: 7933
2022-04-20-16:06:35.573 >> 721.065000 > THCU qorvoRadioClearSrcMatchExtEntry
2022-04-20-16:06:35.589 >> 721.081000 > THCU: DEL DP for nb: 166e0a00:00000002
2022-04-20-16:06:35.605 >> 721.097000 > [DEBG]-MAC-----: SrcAddrMatch - Clearing addr: 166e0a0000000002 -- OK (0)
2022-04-20-16:06:35.620 >> 721.112000 > THCU qorvoRadioAddSrcMatchShortEntry
2022-04-20-16:06:35.620 >> 721.112000 > THCU: ADD DP for nb: d802
```

**Additional context** Add any other context about the problem here.

**Comment by abtink** [comment]
Thanks @thcu-gp for the detailed analysis.

The challenge related to this in `DataPollHandler::HandleSentFrame()` is that we cannot know on parent side whether the child did rx "child ID response" or not,.
- The child may rx the "Child ID response" but parent misses the ack, then child would switch to using short address in data polls.
- The child may not have received the "Child ID response" then it would continue to poll using ext address.

So keeping the src match table as "ext address" on retries from `DataPollHandler` seems reasonable to me. 

That said, I expect that parent to switch to using short address as soon as it sees that child is itself using short address in its data poll. We have this path in the code `in `DataPollHandler` which I expect to kick in:
```c++
void DataPollHandler::HandleDataPoll(Mac::RxFrame &aFrame)
{
    Mac::Address macSource;
    Child *      child;
    uint16_t     indirectMsgCount;

    VerifyOrExit(aFrame.GetSecurityEnabled());
    VerifyOrExit(!Get<Mle::MleRouter>().IsDetached());

    SuccessOrExit(aFrame.GetSrcAddr(macSource));
    child = Get<ChildTable>().FindChild(macSource, Child::kInStateValidOrRestoring);
    VerifyOrExit(child != nullptr);

    child->SetLastHeard(TimerMilli::GetNow());
    child->ResetLinkFailures();
#if OPENTHREAD_CONFIG_MULTI_RADIO
    child->SetLastPollRadioType(aFrame.GetRadioType());
#endif

    indirectMsgCount = child->GetIndirectMessageCount();

    LogInfo("Rx data poll, src:0x%04x, qed_msgs:%d, rss:%d, ack-fp:%d", child->GetRloc16(), indirectMsgCount,
            aFrame.GetRssi(), aFrame.IsAckedWithFramePending());

    if (!aFrame.IsAckedWithFramePending())
    {
        if ((indirectMsgCount > 0) && macSource.IsShort())
        {
            Get<SourceMatchController>().SetSrcMatchAsShort(*child, true);
        }

        ExitNow();
    }
```

Basically if we see on parent that child is sending a data poll using short address, parent would also switch to using short src addr match.

**Comment by thcu-gp** [comment]
So the behaviour does not match:
```
// Enable short source address matching after the first indirect
// message transmission attempt to the child. We intentionally do
// not check for successful tx here to address the scenario where
// the child does receive "Child ID Response" but parent misses the
// 15.4 ack from child. If the "Child ID Response" does not make it
// to the child, then the child will need to send a new "Child ID
// Request" which will cause the parent to switch to using long
// address mode for source address matching.
```

the code:
```
    if (!aFrame.IsAckedWithFramePending())
    {
        if ((indirectMsgCount > 0) && macSource.IsShort())
        {
            Get<SourceMatchController>().SetSrcMatchAsShort(*child, true);
        }

        ExitNow();
    }
```
only kicks in when there is a frame pending, but the child could be polling already with its short address for a long time.
This means that on the first frame pending, there is still an ack with fp=0 (on the extended address), before the switch is made to short address. Then on the next ack the fp can be set to one (which depending on the polling behaviour of a SED could be quiet a long time)

**Comment by thcu-gp** [comment]
Is it an option to first add the short address poll handling (somewhere around indirect_sender.cpp:492) and only remove the extended address handling once a datarequest/data from the short address has been received?

**Comment by abtink** [comment]
The behavior described is in `indirect_sender.cpp` related to this is sort of an optimization/enhancement.
We can technically rely on the mechanism in data poll handler, but as you point out it has a short-coming that it can waste the initial/first data poll (if I recall correctly, the change in `indirect_sender` was added to enahcne this and avoid the waste of data poll but the main mechanism to recover in the case of such failure is still the one in `data_poll_handler`).

-----

> Is it an option to first add the short address poll handling (somewhere around indirect_sender.cpp:492) and only remove the extended address handling once a datarequest/data from the short address has been received?

No I think this model is not used/supported in the code. 

Early on we considered such a model (allow both short and ext addresses to be included src match table at least for some short time), but if I recall correctly there was something tricky about this model for some radio platforms (to be honest I don't remember what it was, maybe it was just its complexity in the code). It seems like a reasonable (and more robust) model to me as a solution, but it requires some re-design of handling of src-match entries.

**Comment by thcu-gp** [comment]
I think the combined short and combined poll handling still is best.
Another option could be to switch over when a poll/data from the short address is received, whether there is a frame pending or not.

The current design might not be entirely compliant with the mac spec:

> In the Frame Control field, the Frame Type subfield shall contain the value that indicates an acknowledgment frame, as shown in Table 79. If the acknowledgment frame is being sent in response to a received data request command, the device sending the acknowledgment frame shall determine whether it has data pending for the recipient. If the device can determine this before sending the acknowledgment frame (see 7.5.6.4.2), it shall set the Frame Pending subfield according to whether there is pending data. Otherwise, the Frame Pending subfield shall be set to one.

At this point when we still expect a poll with an extended address, we reply with fp=0

**Comment by jwhui** [comment]
Closing stale issue.
